File: Data Structures & Algorithms/is-palindrome/submission-3.js
class Solution {
    /**
     * @param {string} s
     * @return {boolean}
     */
    isPalindrome(s) {
         let alphaNumeric = ''
         
        function isAlphaNumeric(c){
            const code = c.charCodeAt(0)

            return (
                (code >= 48 && code <= 57) ||
                (code >= 65 && code <= 90) ||
                (code >= 97 && code <= 122)
            )
        };

        for(const char of s) {
            if(isAlphaNumeric(char)) {
                alphaNumeric += char.toLowerCase();
            }
        }

        let left = 0
        let right = alphaNumeric.length - 1

        while (left < right) {
            if(alphaNumeric[left] === alphaNumeric[right]){
                left++
                right--
            } else {
                return false
            }
        };

        return true
    }
}
//Time Complexity: O(N)
//Space Complexity: O(N)

File: Data Structures & Algorithms/valid-palindrome-ii/submission-3.py
class Solution:
    def validPalindrome(self, s: str) -> bool:
        left = 0
        right = len(s) - 1
        while(left < right):
            if s[left] != s[right]:
                return self.isPalindrome(s, left + 1, right) or self.isPalindrome(s, left, right -1)
            else:
                left += 1
                right -= 1
        return True

    


    def isPalindrome(self, s, l, r):
        while(l < r):
            if s[l] != s[r]:
                return False
            else:
                l +=1
                r -=1 
        return True       